src/Encoder.c

```diff
@@ -4404,14 +4405,14 @@ ZYDIS_EXPORT ZyanStatus ZydisEncoderDecodedInstructionToEncoderRequest(
     }
     request->allowed_encodings = 1 << instruction->encoding;
 
-    if (instruction->operand_count > ZYAN_ARRAY_LENGTH(instruction->operands))
+    if (operand_count > instruction->operand_count)
```

Comment:
I partially agree this check could be dangerous, but at the same time I think it's the best we can come up with. This check was added to the formatter in the exact same manner as well.

Relying on the `instruction.operand_count_visible` is even more dangerous imho. I had the following scenario in my mind:

1. Decode instruction (3 visible operands)
2. Deccode operands. User chooses to only decode 2 operands, `operands` array has a length of 2)
3. Format/Encode -> BAM!

Even if it's kinda redundant and still not perfectly safe, I personally like to always pass the size/length of arbitrary length buffers in public APIs. This way you can directly see the buffer and the length side by side and it should be clear to every user that the provided buffer must be large enough to match the desired count/length.

We can add an additional check for e.g. `ZYDIS_OPERAND_COUNT_MAX`/`ZYDIS_ENCODER_MAX_OPERANDS` on these places, but even then it would access invalid memory in my example above.

Reply:
By "dangerous" I wanted to highlight that new check does something different from the original check. I thought that change in behavior was unintentional. I agree there's no perfect solution here when it comes to handling operands, so `ZydisEncoderDecodedInstructionToEncoderRequest` will be no longer crash-proof as it used to be. I wanted to use `ZydisDecodedInstruction.operand_count_visible` for slightly cleaner API but on the other hand current signature is consistent with other functions that accept "a tuple" of operands and operand count. I don't have a strong opinion about this, so I guess it's ok as it is.

> We can add an additional check for e.g. ZYDIS_OPERAND_COUNT_MAX/ZYDIS_ENCODER_MAX_OPERANDS on these places, but even then it would access invalid memory in my example above.

I would totally bring back check for `ZYDIS_ENCODER_MAX_OPERANDS`. Ok, we cannot prevent user from crashing software with malformed operands but thanks to this check we can at least prevent crashes due to out of bounds accesses to encoder request (that's the intent behind original check).

Reply:
Perfect :-) Will reintroduce the upper bounds check on all places and as well change the `>` to `!=` (this part indeed was unintentional).